stablehlo/reference/Index.h

```diff
@@ -0,0 +1,59 @@
+/* Copyright 2022 The StableHLO Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#ifndef STABLEHLO_REFERENCE_INDEX_H_
+#define STABLEHLO_REFERENCE_INDEX_H_
+
+#include <cstdint>
+#include <vector>
+
+#include "llvm/ADT/ArrayRef.h"
+
+namespace mlir {
+namespace stablehlo {
+
+/// Iterator over the indices of a tensor in major-to-minor order. As an
+/// example, for a tensor with shape [2,3], the iterator enumerates the
+/// indices (0,0), (0,1), (0,2), (1,0), (1,1), and (1,2).
+class IndexSpaceIterator {
+ public:
+  /// \name Constructors
+  explicit IndexSpaceIterator(const std::vector<int64_t> &shape,
+                              bool lastIndex = false);
+
+  /// Get the current index.
+  llvm::ArrayRef<int64_t> operator*() const { return indices_; }
+
+  bool operator==(const IndexSpaceIterator &it) {
+    return shape_ == it.shape_ && counter_ == it.counter_;
+  }
+
+  bool operator!=(const IndexSpaceIterator &it) { return !(*this == it); }
+
+  /// Increment to the next valid index while iterating over the dimensions in
+  /// major-to-minor order.
+  IndexSpaceIterator &operator++();
+  IndexSpaceIterator operator++(int);
+
+ private:
+  std::vector<int64_t> shape_;
+  std::vector<int64_t> indices_;
+  int64_t counter_;
```

Comment:
Do we have to store the counter? This creates opportunities for unexpected overflows. If we need to compare two iterators, we can just compare indices, right?

Reply:
The reason we have the `counter_` is because it also works for the scalar case. 

For the scalar case, both `index_begin` and `index_end` returns iterators with `indices_` member variable set to empty. `it != index_end()` would disallow the the use of the current `indices_` of iterator. 

`counter_` is initially set to 0 and it is supposed to counter till the number of elements which we set to 1 in case of scalars.  `index_begin` sets the `counter_` to 0 and `index_end` sets to to number of elements or 1. `lastIndex` helps us to distinguish the two cases.

I agree that the `lastIndex` might be confusing. Instead I can make all these as arguments to the constructor.
wdyt?

Reply:
Please have a look at the latest upload. Added comments to the member variables. 

In short,
The `index_end` should return an iterator past the end of valid iteration space. The member variable `counter_` helps us achieve that. The member variable `num_element_` help flag increment beyond the last index.

Reply:
In the proposed design, `counter_` and `num_elements_` have to be kept in sync with `shape_`, we have to worry about potential overflows - that's a lot of work.

I think it's possible to model the domain more cleanly - without the need for accompanying class members at all. Here's how it could look like: https://github.com/sdasgup3/stablehlo/pull/1.